packages/agave-renderer/itk_viewer_agave_renderer/renderer.py

```diff
@@ -0,0 +1,116 @@
+class AgaveRendererMemoryRedraw(agave.AgaveRenderer):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def memory_redraw(self):
+        self.cb.add_command("REDRAW")
+        buf = self.cb.make_buffer()
+        # TODO ENSURE CONNECTED
+        self.ws.send(buf, True)
+        #  and then WAIT for render to be completed
+        binarydata = self.ws.wait_for_image()
+        # ready for next frame
+        self.cb = agave.commandbuffer.CommandBuffer()
+        img = PILImage.open(binarydata)
+        rgba = img.convert('RGBA').tobytes()
+        return rgba
+
+# how to handle unknown events
+class UnknownEventAction(Enum):
+    ERROR = auto()
+    WARNING = auto()
+    IGNORE = auto()
+
+class Renderer():
+    def __init__(self, width=500, height=400, unknown_event_action=UnknownEventAction.WARNING):
+        self.width = width
+        self.height = height
+        self.unknown_event_action = unknown_event_action
+
+    async def setup(self):
+        # Note: the agave websocket server needs to be running
+        self.r = AgaveRendererMemoryRedraw()
+        self.r.set_resolution(self.width, self.height)
+
+    async def render(self):
+        r = self.r
+        start_time = time.time()
+        rgba = r.memory_redraw()
+        image_type = ImageType(dimension=2, componentType=IntTypes.UInt8, pixelType=PixelTypes.RGBA, components=4)
+        image = Image(image_type)
+        image.size = [self.width, self.height]
+        image.data = np.frombuffer(rgba, dtype=np.uint8)
+        # lossless
+        # rgba_encoded = encode(image)
+        rgba_encoded = encode(image, not_reverible=True, quantization_step=0.08)
+        elapsed = time.time() - start_time
+        return { "frame": rgba_encoded, "renderTime": elapsed }
+
+    def update_renderer(self, events):
+        r = self.r
+
+        for [event_type, payload] in events:
+           match event_type:
+                case 'density':
+                    r.density(payload)
+                case 'cameraPose':
+                   eye = payload['eye']
+                   r.eye(eye[0], eye[1], eye[2])
+                   up = payload['up']
+                   r.up(up[0], up[1], up[2])
+                   target = payload['target']
+                   r.target(target[0], target[1], target[2])
+                case 'renderIterations':
+                   r.render_iterations(payload)
+                case _:
+                    self.handle_unknown_event(event_type)
+
+    def handle_unknown_event(self, event_type):
+       match self.unknown_event_action:
+            case UnknownEventAction.ERROR:
+                raise Exception(f"Unknown event type: {event_type}")
+            case UnknownEventAction.WARNING:
+                print(f"Unknown event type: {event_type}", flush=True)
+            case UnknownEventAction.IGNORE:
+                pass
+
+    async def start_renderer(self, hypha_server_url, load_image_into_agave_fn, visibility="public"):
```

Comment:
I see this is moved in as a class function.  Kind of liked it as a top level factory func, but no biggie.

Reply:
Yes, I can see it both ways, but since it references `self` quite a bit, I made it a method.

I am renaming it to `connect`, but suggestions are welcome on the naming.